### ISSUE TITLE
`R.compose` should make curried function with correct arity

### ISSUE THREAD

**Issue by anton-rudeshko** [ramda/ramda#736]
Almost same as in https://github.com/ramda/ramda/issues/725

Kinda stange behavior:

```js
R.concat([1], [2]); // [1, 2]
R.concat([1])([2]); // [1, 2]
R.join(',', R.concat([1], [2])); // "1,2"

var concatAndJoin = R.compose(R.join(','), R.concat);
concatAndJoin([1], [2]); // "1,2"
concatAndJoin([1])([2]); // Uncaught TypeError: Cannot read property 'apply' of undefined 
```

Sad =(

------

So I think `R.compose` should return curried function with same arity as it last argument.

```js
R.compose(R.join(','), R.concat); // curried with 2 args, same as R.concat.
```

**Comment by CrossEye** [comment]
Obviously you could call `curry` on this.  And that's been my answer so far.

I've been fighting what I think I'm starting to see as a rearguard action against this.  You can see for instance, https://github.com/ramda/ramda/issues/732#issuecomment-70415923 or #485.  There is a serious issue involved in what to do with functions that don't properly report their arities, either variadic ones or other ones, but I guess I'm starting to be won over to the notion that we're going to have to do this.  I still don't know exactly how to deal with those issues.  But it would be really nice to have the resulting function curried.

Any suggestions for how to handle the variadic or poor-arity-reporting functions?

**Comment by anton-rudeshko** [comment]
> Obviously you could call curry on this. And that's been my answer so far.

I want to insist on Ramda principle listed everywhere:

> Ramda functions are automatically curried. 

And fundamental to FP `compose` isn't curried? Sounds like a joke to me.

Yes, obviously I can `curry` by hand. But what's the point?

I think that without common approach for currying variadic functions in JS (or in Ramda in particular) we need to leave this decision to users because solution may depend strongly on use case.

**Comment by CrossEye** [comment]
> I want to insist on Ramda principle listed everywhere:

>> Ramda functions are automatically curried.

That's usually preceded with "almost all". :smile:  There is a reason for this.  We have a few variadic functions, which cannot comfortably be curried.  But that's different from saying that functions you create using Ramda will be curried as well.  I'm starting to come to think that we should go that route as much as possible, but there are some complexities to work through.


> And fundamental to FP `compose` isn't curried? Sounds like a joke to me.

Actually, that's one of our variadic functions.  So too it's order-reversed counterpart `pipe`.  We want people to be able to write long  lists of functions in a single pipeline, as in this snippet of our build script:

```javascript
    R.pipe(
        identifierToFilename,
        parse,
        R.prop('ast'),
        R.prop('body'),
        R.takeWhile(R.where({type: 'VariableDeclaration', kind: 'var'})),
        R.pluck('declarations'),
        R.map(R.tap($pathEq('length', 1))),
        R.map(R.head),
        R.map(R.tap($pathEq('init.type', 'CallExpression'))),
        // .... many more
    );
```

Because `compose` and `pipe` are variadic, they will probably never themselves be curried, but I think that's a different question than the one at issue here.

> I think that without common approach for currying variadic functions in JS (or in Ramda in particular) we need to leave this decision to users because solution may depend strongly on use case.

It's not a question of currying variadic functions.  (I don't think there's a reasonable way to do that within the constraints we've imposed on ourselves.)  The question is about the behavior of `compose`/`pipe`, and how that behavior depends on the first function to be executed.

Whether or not the function is variadic doesn't matter so much.  This would work fine if the result of `compose` were curried:

```javascript
var stats = compose(
    converge(
        mixin,
        pipe(min, createMapEntry("min")),
        pipe(max, createMapEntry("max"))
    ),
    unapply(identity)
);
stats(8, 6, 7, 5, 3, 0, 9); //=> {min: 0, max: 9}
```

But if the function has optional parameters, you'd run into real issues

```javascript
var fullName = function(first, last, suffix) {
    return first + ' ' + last + (suffix ? ', ' + suffix : '');
};
```

Using our current compose, we're fine:

```javascript
var f1 = compose(flip(concat)(' played a monster.'), fullName);
f1('Lon', 'Chaney', 'Jr'); //=> 'Lon Chaney, Jr played a monster.'
f1('Boris', 'Karloff'); //=> 'Boris Karloff played a monster.'
```

But using an auto-currying compose we run into this problem:

```javascript
var f2 = compose(flip(concat)(' played a monster.'), fullName);
f2('Lon', 'Chaney', 'Jr'); //=> 'Lon Chaney, Jr played a monster.'
f1('Boris', 'Karloff'); //=> function :: String -> String
```

And that's the reason for my hesitation.  Users cannot compose functions with named optional parameters as the first function in a composition if we auto-curry the composed result.  It might be a rare enough case that we want to offer an auto-curried version as our main API and one without under some alternative name, but I don't think we can go blindly into this.

**Comment by stoeffel** [comment]
I am with @CrossEye on this (despite my PR). To lose the possibility of using compose with functions with optional parameters is the bigger loss than having a curried function returned.

**Comment by aaditmshah** [comment]
@stoeffel I am actually against @CrossEye on this. I have stated time and again that I am [against variadic functions in JavaScript](http://stackoverflow.com/a/26701110/783743) for the following reasons:

1. Variadic functions are unprincipled. They add unnecessary complexity to programs. The `length` of variadic functions is always incorrect (which makes Ramda a hypocritical library because otherwise it always tries to correctly report `length` of all functions). In the absence of variadic functions you can make strong guarantees about the `length` of your functions.
2. [Variadic functions and curried functions don't mix](https://twitter.com/robotlolita/status/556972021108969473). They are like oil and water. If you want to write consistent code then you could either use variadic functions or curried functions, but not both. Hence, Ramda is inconsistent. I believe that it is sacrilege for a functional programming library to export variadic functions.
3. Variadic functions are unnecessary. Every variadic function in JavaScript can be converted into a non-variadic function which accepts an array of &ldquo;variadic arguments&rdquo;. The only difference would be that the non-variadic function call would have two extra characters (i.e. an open square bracket and a close square bracket, for arrays). Two extra characters is a very small price to pay for simplicity and explicitness. As the [Zen of Python](https://www.python.org/dev/peps/pep-0020/) states, &ldquo;Explicit is better than implicit. Simple is better than complex.&rdquo; BTW, if you want to use `slice` or `concat` on the `arguments` object then you need to convert it into an array anyway.
4. Using the `arguments` object not only makes the code slower, but also is an [optimization killer](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments) when used incorrectly in V8 (which means that Ramda code in node is mostly interpreted instead of JIT compiled). Reading the Ramda source code, I see many instances where performance could have been better if they had simply not used the `arguments` object. Non-variadic functions don't need to worry about this because they don't have to deal with the `arguments` object anyway. The only safe things you can do with the `arguments` object is access its properties and pass it to `apply`. Beside that, every other operation on the `arguments` object is an optimization killer, including passing it to other functions.

So what's the solution?

I usually use the following `chain` function for composing a chain of functions:

```javascript
var chain = curry(function (a, x) {
    var length = a.length;
    while (length > 0) x = a[--length](x);
    return x;
});
```

It can be used as follows:

```javascript
var foo = chain([map(inc), filter(odd), take(5)]);

foo([1,2,3,4,5,6,7,8,9,10]); // [2,4,6]
```

Which is equivalent to:

```haskell
let foo = map (+1) . filter odd . take 5

foo [1,2,3,4,5,6,7,8,9,10]
```

It also allows you to write code like:

```javascript
chain([map(inc), filter(odd), take(5)], [1,2,3,4,5,6,7,8,9,10]); // [2,4,6]
```

Which is equivalent to:

```haskell
map (+1) . filter odd . take 5 $ [1,2,3,4,5,6,7,8,9,10]
```

Simple and elegant, all because we didn't use variadic functions.

Other concerns:

1. The `pipe` and `compose` functions should return curried functions. If `curry` itself curries the result of functions which return functions, then this problem will be solved without having to change the `pipe` and `compose` functions at all. See https://github.com/ramda/ramda/issues/732.
2. Optional named parameters. This is a design principle. I believe that no named parameter should ever be optional. If you want optional parameters, then either create a separate function which calls the main function with the default value of the optional argument, or else pass in a non-optional &ldquo;options&rdquo; object.

In conclusion, I understand that you are trying to be &ldquo;good neighbours&rdquo; by supporting lots of ancillary design concerns. However, I believe that you should not do so. Don't worry about appeasing non-functional JavaScript programmers. If you're aiming to be a functional programming library then you should have functional design constraints. Do not try to appease the general JavaScript crowd because JavaScript is a soup, and it's not possible to have a one-library-that-fits-all-purposes. Just concentrate on making Ramda __functional__, not imperative.

**Comment by CrossEye** [comment]
@aaditmshah:

There's a lot to respond to here, and it's almost time for me to go to bed, so I may not get to everything.  The first thing I want to say is thank you very much for your very interesting contributions to the discussions here.  Several of your suggestions have either been tried or are still being contemplated.

My general reaction to your suggestions on these topics, though, is that what you're trying to do is somewhat different from what Ramda is trying to do.  Ramda is noting where functional programming overlaps well with Javascript, and where it can be made to do so with a little help, and then fills in the small gaps to make it easier.  The chief thing it is meant to support is building simple functional pipelines out of small curried functions.  It seems to me that you have a more ambitious goal of pushing further, of making Javascript more closely resemble typical functional languages like Haskell.

Nothing demonstrates this difference more clearly than your reaction to variadic functions.

> 1. Variadic functions are unprincipled.

I disagree.  They don't match well with strong typing, that's for certain, but that only implies that their principles disagree with those of the strongly typed world.  I would also argue against any notion of hypocrisy on Ramda's part due variadic functions.  Ramda does what it can to properly report the arity of `composed` and `curried` functions.  When the functions that we work with do not properly report their own arity, we cannot do better than they do.  Hypocrisy is something else.


> 2. Variadic functions and curried functions don't mix.

This is misspoken.  Ramda mixes them without any serious issues.  What is true is that you cannot do is to curry a variadic function.  That makes no sense.  That's one of the reasons that Ramda has very few variadic functions.  But it does have a few.  They could easily be rewritten to take lists of arguments rather than get their values from the `arguments` object, but that gains nothing I can see except to comply with an abstract, external notion of purity.


> 3. Variadic functions are unnecessary.

This is true, of course.  But it's besides the point.  Ramda is not in the business of dictating how users write their code.  In fact, for the most part, functions returned from Ramda can even be used as object methods if that's how the user wants to employ them.  (By this I mean that the `this` context is maintained.  This is not a major goal, but it's nice that we can usually accomplish this, and so when we notice that it's broken we fix it.)  So, yes, we are writing the library to enable a certain style of programming, but that doesn't mean that we are trying to enforce that style on our users.  If they want to start a composition off with a variadic function, that will work just fine.


> 4.

**Comment by ashnur** [comment]
Optional arguments to me seem silly. I either pass a list or have more functions. If someone is having optional arguments then they need IF-ELSE logic in their function, which in most cases are simple defaults, but it's just a question of time when it will get out of hand. That said, if someone wants to write silly code, I am in no position to tell them to stop. 

Of course, if one gets to know Ramda well and accepts its way of how it works and its limitations, one can avoid those situations where these problems arise. In my view this unnecessarily steepens the learning curve which, for most js programmers (like me for example), with our imperative customs, is quite steep enough already. I feel like these problems hit only inexperienced FP programmers. Someone who is experienced with FP and wants to use Ramda will probably see and avoid them anyway. I think there should be a more flexible, more forgiving way to solve this issue.

**Comment by CrossEye** [comment]
> Someone who is experienced with FP and wants to use Ramda will probably see and avoid [`compose`/`curry` problems] anyway. I think there should be a more flexible, more forgiving way to solve this issue.

I think there is.   I believe our best bet is to change our default `compose` to curry before returning and add a second version that does not do so.   I can't think of a good name and that worries me.  Documentation won't be fun either.   But I can't think of anything better.

**Comment by anton-rudeshko** [comment]
Sorry, you're going too fast and I can't follow that quick =) 
I need to make some time to read all your comments...

**Comment by anton-rudeshko** [comment]
> > And fundamental to FP compose isn't curried?

> Actually, that's one of our variadic functions.

I put it wrong. I meant that result of `compose` isn't curried, not `compose` itself. 

> It might be a rare enough case that we want to offer an auto-curried version as our main API and one without under some alternative name

***+1*** here.

It looks to me that @aaditmshah has very strong arguments.

**Comment by CrossEye** [comment]
I just noticed that I'd accidentally closed this issue.  It certainly wasn't intentional.  I think this is a very important discussion.  Reopening.

**Comment by buzzdecafe** [comment]
looks like we might be heading for `compose/composeN` split, where the former curries based on initial function length and curries the output function to that, and the latter takes an arity `n` and curries the output function to that. (So `compose` could be implemented as `composeN(lastFn.length, fns...)`)

**Comment by davidchambers** [comment]
Resolved by #1260:

```javascript
> R.compose(R.join(','), R.concat).length
2
> R.compose(R.join(','), R.concat)(['foo', 'bar'], ['baz'])
'foo,bar,baz'
> R.compose(R.join(','), R.concat)(['foo', 'bar'])(['baz'])
'foo,bar,baz'
```

**Comment by dalgard** [comment]
I see that the function returned by `compose` and `pipe` has correct arity – the same as the first function. However, it still isn't curried as expected.

**Comment by CrossEye** [comment]
Yes,  doing this caused serious issues. 

You can see the discussion around it in #1318 and the change back from #1260 in #1391.

**Comment by dalgard** [comment]
It's a shame. I hope to see this revived in the future.

**Comment by CrossEye** [comment]
> It's a shame. I hope to see this revived in the future.

And I hope to see it die a painful death, all alone, abandoned and despondent.  :-)

David's [Alice and Bob story][ab] demonstrated that we cannot satisfy all reasonable requirements simultaneously.  I argue that the current behavior is the most useful compromise.


  [ab]: https://github.com/ramda/ramda/issues/1318#issuecomment-139103766
